Add Granite Speech Support

alex-jw-brooks opened the pull request:
# What does this PR do?

This PR adds support for (upcoming) Granite Speech models. In terms of model architecture, it uses a conformer-based encoder, with a blip2 qformer-based projector to encode the audio, and masks it into a granite llm. This model also uses an audio-specific lora adapter, which should _only_ be enabled when the model is processing audio inputs.

Currently this is handled by using the transformers/peft mixin to load the adapter from the same directory as the base model and overriding `.generate` to turn the lora on/off based on the presence of audio input features before forwarding to `.generate` on the superclass. Doing it this way, as opposed to encapsulating a Peft Model, makes this a lot cleaner in a lot of ways and keeps the lora out of the modeling code.

(This PR is a collaboration with @avihu111  and @gsaon)
## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

alex-jw-brooks committed:
First pass at speech granite

Add encoder / projector, rename things

alex-jw-brooks committed:
Combine into one model file with causal lm outputs for forward

alex-jw-brooks committed:
Add loss calc

alex-jw-brooks committed:
Fix config loading

Signed-off-by: Alex-Brooks <Alex.brooks@ibm.com>

alex-jw-brooks committed:
Split new / old loading logic

alex-jw-brooks committed:
Use transformers integration for loading peft adapters

alex-jw-brooks committed:
Add generation wrapper for selective lora enablement

alex-jw-brooks committed:
Add note for qformer encoder automodel

alex-jw-brooks committed:
Guard torch/audio imports in feature extractor

alex-jw-brooks committed:
Handle granite speech autoclasses

alex-jw-brooks committed:
Handle optional deps in package structure for granite speech

alex-jw-brooks committed:
Add granite pretrained model def for init

alex-jw-brooks committed:
Add dummy objects for torch/torchaudio

alex-jw-brooks committed:
Add tests for granite speech processor

alex-jw-brooks committed:
Minor formatting fixes and refactoring

alex-jw-brooks committed:
Add options for falling back to config in forward

alex-jw-brooks committed:
Tentative model docstrings for granite speech

alex-jw-brooks committed:
Fix config type

alex-jw-brooks committed:
Remove legacy load

alex-jw-brooks committed:
Allow non-lora variants for granite speech

alex-jw-brooks committed:
Override weight tying for llm

alex-jw-brooks committed:
Use text config instead of llm config

alex-jw-brooks committed:
Add output embeddings getter to fix weight tying

alex-jw-brooks committed:
Fix relative imports

avihu111 committed:
computing the number of audio features, based on the raw audio sequence.

avihu111 committed:
collating audio inputs, and keeping the original lengths.

avihu111 committed:
asserted we have text. otherwise we can't specify the audio special token.

avihu111 committed:
assering the number of audio-symbols/audios match correctly.
running get validated_audios only when audio is present

avihu111 committed:
indentation bugfix + supporting different feature lengths when expanding audio.

avihu111 committed:
redundant, done in _get_validated_text

avihu111 committed:
adapting the tests:
- we must have text (not either audio or text)
- _get_num_audio_features takes a list of raw lengths, provided it insetad.

alex-jw-brooks committed:
Minor cleanup, remove unused import

alex-jw-brooks committed:
Add more tests for batch feature processing

alex-jw-brooks committed:
Allow setting offset in rel position embeddings

alex-jw-brooks committed:
Add config option for warning if peft is not installed w/ lora

Rocketknight1 commented:
cc @eustlb

alex-jw-brooks committed:
Port blip2 qformer code into granite speech

alex-jw-brooks committed:
Add sad test for numpy arr processing

alex-jw-brooks committed:
Allow numpy arrays / tuples in granite speech processor

alex-jw-brooks committed:
Fix config type for projector

avihu111 committed:
- pad instead of creating a zeros tensor, to keep the original dtype/device (support bfloat16)
- cast input_features to the model dtype (support bfloat16)

avihu111 committed:
merge Blip2QFormerConfig to GraniteSpeechProjectorConfig

avihu111 committed:
prevent a crash when re-saving/loading the model (line 109)

avihu111 committed:
consider additional edge cases during preprocessing.

avihu111 committed:
consider additional edge cases during preprocessing.

avihu111 committed:
add features mask for batched inference (bugfix)

alex-jw-brooks committed:
Add set input/output embeddings for granite speech

alex-jw-brooks committed:
Fix feature dim check in processor test

alex-jw-brooks committed:
Pop input features in embed test for granite speech

alex-jw-brooks committed:
Small fixes for test edge cases

Add granite speech to seq2seq causal lm mapping names

alex-jw-brooks committed:
Add small tests for granite speech model

alex-jw-brooks committed:
Minor refactor, remove multiaudio processor tests

alex-jw-brooks committed:
Fix data parallelism test

alex-jw-brooks committed:
Standardize model class names

alex-jw-brooks committed:
Fix check for copies

alex-jw-brooks committed:
Fix misaligned init check

alex-jw-brooks committed:
Skip granite speech in checkpoint check

alex-jw-brooks committed:
Use default for tie_word_embeddings in granite speech

alex-jw-brooks committed:
Fix non documentation granite speech repo issues

alex-jw-brooks committed:
Fix comments and docstring checks

alex-jw-brooks committed:
Add placeholder docs for granite speech

alex-jw-brooks committed:
Fix test naming collision

alex-jw-brooks committed:
Code formatting

alex-jw-brooks committed:
Rerun torch dummy obj regen

alex-jw-brooks committed:
Fix save pretrained for granite speech

alex-jw-brooks committed:
Import sorting

alex-jw-brooks committed:
Fix tests typo

alex-jw-brooks committed:
Remove offset hack

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+    def set_input_embeddings(self, value):
+        self.embeddings.word_embeddings = value
+
+    # Copied from transformers.models.blip_2.modeling_blip_2.Blip2QFormerModel._prune_heads
+    def _prune_heads(self, heads_to_prune):
+        """
+        Prunes heads of the model. heads_to_prune: dict of {layer_num: list of heads to prune in this layer} See base
+        class PreTrainedModel
+        """
+        for layer, heads in heads_to_prune.items():
+            self.encoder.layer[layer].attention.prune_heads(heads)
+
```
to remove I don't think we need this

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+                # add cross attentions if we output attention weights
+                outputs = outputs + cross_attention_outputs[1:-1]
+
+            layer_output = apply_chunking_to_forward(
+                self.feed_forward_chunk_query,
+                self.chunk_size_feed_forward,
+                self.seq_len_dim,
+                query_attention_output,
+            )
+
```
not a fan of calling a function, with arg being a function that needs self.xxxx

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+        else:
+            self.has_cross_attention = False
+
+        if config.use_qformer_text_input:
```
is this true of false for the released model?

avihu111 replied:
Good point. It's always set to False, so we can remove this part to simplify the code.

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+
+
+# Copied from transformers.models.blip_2.modeling_blip_2.Blip2QFormerAttention with Blip2->GraniteSpeech
+class GraniteSpeechQFormerAttention(nn.Module):
```
I would just use on e of the recent attention class like Llama, here the output is split from the attention itself, which is not really common anymore!

ArthurZucker replied:
for cross attention Whisper is a great source of truth that has cross attn

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+
+
+# Copied from transformers.models.blip_2.modeling_blip_2.Blip2QFormerMultiHeadAttention with Blip2->GraniteSpeech
+class GraniteSpeechQFormerMultiHeadAttention(nn.Module):
```
let's not use Blip2! we should not register hooks and save attentions ! Let's rather take llama or whisper as references!

ArthurZucker commented on src/transformers/models/granite_speech/feature_extraction_granite_speech.py:
```diff
@@ -0,0 +1,118 @@
+        if self.melspec is None:
+            self.melspec = torchaudio.transforms.MelSpectrogram(**self.melspec_kwargs)
+
+    def __call__(
```
not sure this supports the batch features? Also not super aligned with our API here we use `return_tensors=...` and etc. Have a look at PHI4 feature extraction code to make it more aligned!

avihu111 replied:
We support batched feature extraction and inference, but the batching logic is done in our Processor.
I'll re-organize this to match existing feature-extractors like Phi4. Thanks!

avihu111 replied:
I attempted to make the feature_extraction/preprocessor resemble the logic/namings used in Phi4.
I moved all the audio validating+batching+attention_mask logic to the `audio_processor`.

ArthurZucker replied:
Nice!!!

ArthurZucker commented on src/transformers/models/granite_speech/__init__.py:
```diff
@@ -0,0 +1,88 @@
+# Copyright 2025 EleutherAI and The HuggingFace Inc. team. All rights reserved.
```
should use the new init format! Available for example for Llama's __init__.py 🤗

ArthurZucker commented on src/transformers/models/granite_speech/processing_granite_speech.py:
```diff
@@ -0,0 +1,150 @@
+        text_inputs = self.tokenizer(text, padding=True, **kwargs)
+        return BatchFeature(data={**text_inputs, **speech_inputs})
+
+    def _expand_audio_placeholders(self, text: list[str], num_audio_features: List[int]):
```
let's align with how we do this in `vision language models`! (mostly just naming) I have not checked the qwen audio processor, but we should do something similar cc @eustlb

ArthurZucker commented on tests/models/granite_speech/test_modeling_granite_speech.py:
we need an integration test here!!!

alex-jw-brooks replied:
Added! Currently it's disabled while we await the model release, but everything passes on the version I have.

Will submit a PR to turn the tests on and add an example to docs once the model is out!

ArthurZucker reviewed (commented):
Sounds great!!!

alex-jw-brooks committed:
Pass args through encoder config

alex-jw-brooks committed:
Remove unused prune heads from blip2

avihu111 committed:
removing einsum. replaced with explicit multiplication (relative positional encodings) and sdpa attention.

avihu111 committed:
remove Sequential from ConformerFeedForward and ConformerConvModule. + fix for sdpa attention

avihu111 committed:
remove GraniteSpeechConformerScale

avihu111 committed:
rename to hidden_states

avihu111 committed:
rename conformer layers to self.layers, remove the first linear from the list to keep the list homogenous.

avihu111 committed:
move pre-norm to the attention/feedforward blocks (avoid complex module wrapping)

avihu111 committed:
adding pre_norm into forward

avihu111 committed:
feature extractor refactoring to resemble how it's done in phi4multimodal.

avihu111 committed:
rename feature_extractor to audio_processor

avihu111 commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+        q, k, v = [t.reshape(bs, nb, context_size, h, -1).transpose(2, 3) for t in (q, k, v)]
+
+        dots = einsum("b m h i d, b m h j d -> b m h i j", q, k) * self.scale
+
```
Done! Replaced this `einsum` with sdpa attention and the second (relative positions dot product) with a standard dot product.

avihu111 commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+        super().__init__()
+        self.net = nn.Sequential(
+            nn.Linear(dim, dim * mult), nn.SiLU(), nn.Dropout(dropout), nn.Linear(dim * mult, dim), nn.Dropout(dropout)
+        )
```
agreed and done, much cleaner this way, and tensor names are readable.

avihu111 commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+            nn.Conv1d(inner_dim, dim, 1),
+            GraniteSpeechConformerPermute(dims=(0, 2, 1)),
+            nn.Dropout(dropout),
+        )
```
Done.
I also removed the permute layer and dropped the `not causal` part of the logic which is never used.

avihu111 commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,1393 @@
+        x = self.conv(x) + x
+        x = self.ff2(x) + x
+        x = self.post_norm(x)
+        return x
```
Done :)

avihu111 committed:
bugfix: input_feature_mask fix to get the exact number tokens.

alex-jw-brooks committed:
Fix pytest decorator in processor test

alex-jw-brooks committed:
Add (disabled) integration tests for granite speech

ArthurZucker reviewed (commented):
Have not reviewed all yet but: 
> The current blip2 qformer code does work without any issues though. I believe the reason we did this instead of abstracting it and loading it out of the config like other models, e.g., the visual encoder in llava next ([here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava_next/modeling_llava_next.py#L362)) is that Blip2QFormerModel can't currently be created through AutoModel.from_config, since it's not an attribute on transformers.models, and it seemed like an anti pattern to add it there without also pulling it out into its own model dir 😓

A way to fix this is just to make the qformer model part of the public API! 
Something along the lines of #36493 should be possible! 

Also let's make sure you create a `modular_granite_speech.py` model, this will help keep the model up to date and isolate differences / simplify the actual diff! 🤗 https://huggingface.co/docs/transformers/en/modular_transformers for details and don't hesitate to ping me!

alex-jw-brooks committed:
Fix handling of optional feature masking

alex-jw-brooks committed:
Loosen validation in processing for vLLM compatability

alex-jw-brooks committed:
Formatting fixes

alex-jw-brooks committed:
Update init structure to mirror llama

alex-jw-brooks committed:
Make granite speech projector generic

alex-jw-brooks committed:
Update test config to reflect generic projector

alex-jw-brooks committed:
Formatting fixes

alex-jw-brooks committed:
Fix typos, add license

alex-jw-brooks committed:
Fix undefined var in input processing

alex-jw-brooks committed:
Cleanup and expose ctc encoder

alex-jw-brooks committed:
Add missing config docstrings

alex-jw-brooks committed:
Better var names, type hints, etc

alex-jw-brooks committed:
Set attn context size in init

alex-jw-brooks committed:
Add max pos emb to encoder config

alex-jw-brooks committed:
Cleanup feature extractor

alex-jw-brooks committed:
Add granite speech architecture details

alex-jw-brooks commented:
Hey @ArthurZucker! I think we have addressed all of the changes you have requested, except for the `modular_granite_speech` part. Could you please take a look when you get the chance?

Regarding adding a modular file - I've opened a PR to expose `Blip2QFormer` here (and also rebased this PR on top of it). With this change, we ca delete the mountain of copied Blip2 code and generically load the projector through `AutoModel.from_config`. similar to the vision tower in most VLMs 🎉 There is no functional diff with the existing Blip2 QFormer code, and it was only originally copied because it wasn't registered as a special auto model yet.

Given this, the model code is pretty much independent from existing `transformers` modules (i.e., not copying or inheriting from anything), so the code in the modular file would be identical to the current modeling code. Any thoughts on how to proceed here would be super appreciated!

alex-jw-brooks committed:
Remove granite speech qformer ref

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+        return hidden_states
+
+
+class GraniteSpeechConformerFeedForward(nn.Module):
```
this should be above! 🤗

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+        return hidden_states
+
+
+class GraniteSpeechConformerBlock(nn.Module):
```
definition of clases being used should be placed above!

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+
+
+class GraniteSpeechConformerAttention(nn.Module):
+    """Attention for conformer blocks with shaw's relpos embeddings."""
```
can we have a link to paper! 🤗

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+
+        query_states = self.to_q(hidden_states)
+        key_states, value_states = self.to_kv(hidden_states).chunk(2, dim=-1)
+        query_states, key_states, value_states = [
```
let's have this written explicitly please

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+        ]
+
+        # shaw's relative positional embedding
+        seq = torch.arange(self.context_size, device=hidden_states.device)
```
we should comput this in the `Model` this way each layer receives it already preopaerd no?

alex-jw-brooks replied:
Good point! I moved the calculation for the clamped distances to `__init__` for this class - it duplicates it per block, but keeps the init/forward interfaces clean feeling and is not a very big tensor (200, 200 for the next model) 🙂

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+            mask_value = -torch.finfo(pos_attn.dtype).max
+            pos_attn[:, -1, :].masked_fill_(mask, mask_value)
+
+        with torch.nn.attention.sdpa_kernel(torch.nn.attention.SDPBackend.MATH):
```
we usually try to have the `attention_interface` but should be alright

alex-jw-brooks replied:
Sounds good! Happy to make a potential follow-up PR with that in the future 😄

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+    def __init__(self, config: GraniteSpeechEncoderConfig):
+        super().__init__()
+        inner_dim = config.hidden_dim * config.conv_expansion_factor
+        padding = self.calc_same_padding(config.conv_kernel_size)
```
lets remove the function entirely!

ArthurZucker commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,671 @@
+        special_audio_mask = is_audio_index.unsqueeze(-1)
+        audio_features = audio_features.to(inputs_embeds.device, inputs_embeds.dtype)
+        if input_features_mask is not None:
+            assert torch.all(is_audio_index.int().sum(dim=1) == input_features_mask.int().sum(dim=1)).item(), (
```
let gracefully raise an error!

ArthurZucker reviewed (approved):
Very nice! 🤗 Thanks for bearing with me!

alex-jw-brooks committed:
Add paper link, explicit calc for qkv

alex-jw-brooks committed:
Calculate padding directly in depthwise conv1d init

alex-jw-brooks committed:
Raise value error instead of asserting

alex-jw-brooks committed:
Reorder class defs (classes used at top)

alex-jw-brooks committed:
Precompute relpos distances

alex-jw-brooks committed:
Run formatting

alex-jw-brooks commented:
Awesome, thank you for all the guidance @ArthurZucker! I made the changes and have verified that the skipped tests still look good on our side 🙂

Will open a follow-up PR once the version of the model that is happy with transformers is up on the hub to add an example to the docs and enable the tests!

alex-jw-brooks committed:
Pass attention distances through forward

eustlb commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,677 @@
+            if input_features.dtype != self.dtype:
+                input_features = input_features.to(self.dtype)
+            # Get the audio features from the encoder / projector
+            audio_features = self.get_audio_features(input_features)
```
better naming (they live in the embeddings space)
```suggestion
            audio_embeds = self.get_audio_features(input_features)
```

eustlb commented on src/transformers/models/granite_speech/modeling_granite_speech.py:
```diff
@@ -0,0 +1,677 @@
+            audio_features = self.get_audio_features(input_features)
+
+            # Merge the audio features into the LLM embeddings
+            inputs_embeds = self.get_merged_audio_embeddings(
+                input_ids=input_ids,
+                audio_features=audio_features,
+                input_features_mask=input_features_mask,
+            )
```
to modify to that it really takes `input_features_mask` (so having `input_features_mask` the same shape as `input_features`)

eustlb reviewed (approved):
Great work guys !! 🤗
Mostly comments about the processor part that can be a bit improved. Do not hesitate to take inspiration from `processing_mllama.py`, `processing_llava.py`(the API for audio should be the same as for vision) and `processing_qwen2_audio.py`.

alex-jw-brooks committed:
Apply suggestions from code review

Co-authored-by: eustlb <94853470+eustlb@users.noreply.github.com>

alex-jw-brooks committed:
Add todo for using common batch feature extraction

alex-jw-brooks committed:
Rename audios/features

alex-jw-brooks committed:
Ensure chat template may be provided to processor

alex-jw-brooks committed:
Move granite speech docs to audio models

alex-jw-brooks committed:
Add todos for input proc refactoring

alex-jw-brooks commented on src/transformers/models/granite_speech/feature_extraction_granite_speech.py:
```diff
@@ -0,0 +1,191 @@
+        transform is not JSON serializable.
+        """
+        if self.melspec is None:
+            self.melspec = torchaudio.transforms.MelSpectrogram(**self.melspec_kwargs)
```
That sounds great! I added a TODO here with both of our names, but please feel free to tag me as the audio utils are developed, I'm happy to try to help change this to use the common utilities 😄

alex-jw-brooks commented on tests/models/granite_speech/test_modeling_granite_speech.py:
```diff
@@ -0,0 +1,393 @@
+        cleanup(torch_device, gc_collect=True)
+
+    def _get_prompt(self, tokenizer):
+        chat = [
+            {
+                "role": "system",
+                "content": "Knowledge Cutoff Date: April 2024.\nToday's Date: December 19, 2024.\nYou are Granite, developed by IBM. You are a helpful AI assistant",
+            },
+            {
+                "role": "user",
+                "content": "<|audio|>can you transcribe the speech into a written format?",
+            },
+        ]
```
Sounds good! We will submit an example once the model that is compatible with transformers is released and use this API! Still working on the final processor chat template from our side 🙂

alex-jw-brooks committed:
Fix import order

alex-jw-brooks committed:
Guard torch import

alex-jw-brooks committed:
Use relative imports

alex-jw-brooks committed:
Require torch backend for processor in granite speech

alex-jw-brooks committed:
Add backend guards in feature extractor

alex-jw-brooks commented on docs/source/en/_toctree.yml:
```diff
@@ -485,6 +485,8 @@
       - local: model_doc/granitemoeshared
         title: GraniteMoeShared
+      - local: model_doc/granite_speech
+        title: GraniteSpeech
```
Good point! Moved 🙂

eustlb committed:
Merge branch 'main' into granite_speech
